lib/cicognara/bulk_entry_indexer.rb

```diff
@@ -1,16 +1,113 @@
 module Cicognara
+  class EntryIndexer
+    attr_reader :entry
+    def initialize(entry, index, cache, books_cache)
+      @entry = entry
+      @index = index
+      @cache = cache
+      @books_cache = books_cache
+    end
+
+    def books
+      @books ||= @entry.books.uniq(&:id)
+    end
+
+    def books_cache_key
+      return if books.empty?
+      books.map(&:id).sort.map(&:to_s).join('_')
+    end
+
+    def book_indexer
+      @book_indexer ||= BookIndexer.new(books)
+    end
+
+    def book_documents
+      return {} if books.empty?
+      # return @books_cache[books_cache_key] if @books_cache.key?(books_cache_key)
+      @book_documents = book_indexer.to_solr
+      @books_cache[books_cache_key] = @book_documents
+      @books_cache[books_cache_key]
+      @book_documents
+    end
+
+    def book_document_values
+      # return [] if book_documents.values.nil?
+      book_documents.values.compact
+    end
+
+    def entry_cache_key
+      entry.id.to_s
+    end
+
+    def cache
+      # return @cache[entry_cache_key] if @cache.key?(entry_cache_key)
+      book_documents
+      @cache[entry_cache_key] = WithBookCache.new(@entry, book_documents)
+      @cache[entry_cache_key]
+      WithBookCache.new(@entry, book_documents)
+    end
+
+    def documents
+      @documents ||= cache.to_solr
+    end
+
+    def index
+      @index.add(documents)
+      @index.commit
+    end
+  end
+
   class BulkEntryIndexer
     attr_reader :entries
-    def initialize(entries)
+    def initialize(entries, index)
       @entries = entries
+      @index = index
+      @entries_cache = {}
+      @books_cache = {}
+    end
+
+    def entry_indexers
+      @entry_indexers ||= @entries.map { |entry| EntryIndexer.new(entry, @index, @entries_cache, @books_cache) }
```

Comment:
MarcRecord objects (retrieved from Book objects linked to Entry objects) have the full XML documents accessible using `#source` here